File: valofosho/BOJ/python/boj1774.py
"""
BOJ1774 - 우주신과의 교감

로직 정의
1. 혼자만 떨어진 애들을 대상
2. 계속해서 Union-Find 로 떨어진 애들 찾기
3. 모든 거리 계산한 뒤 짧은 통로부터 시도
4. Union-Find로 연결

"""

import sys
input = sys.stdin.readline

N, M = map(int, input().split())
maps = [tuple(map(float, input().split())) for _ in range(N)]
parent = [i for i in range(N)]

def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])
    return parent[x]

def union(a, b):
    a, b = find(a), find(b)
    # 루트가 다른 애가 나오면 더해버려
    if a != b:
        parent[b] = a

for _ in range(M):
    a, b = map(int, input().split())
    # 선 연결은 1-based 이므로 -1 연산 후 union
    union(a-1, b-1)

edges = []
# a모든 정점에 대해 L2거리 edge에 넣기
for i in range(N):
    for j in range(i+1, N):
        x1, y1 = maps[i]
        x2, y2 = maps[j]
        dist = ((x1-x2)**2 + (y1-y2)**2)**0.5
        edges.append((dist, i, j))
# 크루스칼을 위해 dist 기준 오름차순 정렬
edges.sort()

total = 0
for dist, a, b in edges:
    # a -> b가 이미 연결되어 있지 않다면
    if find(a) != find(b):
        union(a, b)
        # 연결하고, 거리 추가
        total += dist
print(f"{total:.2f}")

File: bigsky/BOJ/Java/boj1774.java
// BOJ1774(D3): 우주신과의 교감
import java.io.*;
import java.util.*;

public class boj1774 {
    static int[] parent;
    static Point[] points;

    static class Point {
        int x, y;
        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static class Edge implements Comparable<Edge> {
        int u, v;
        double cost;

        Edge(int u, int v, double cost) {
            this.u = u;
            this.v = v;
            this.cost = cost;
        }

        @Override
        public int compareTo(Edge o) {
            return Double.compare(this.cost, o.cost);
        }
    }

    static int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    static boolean union(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return false;
        parent[y] = x;
        return true;
    }

    static double distance(Point p1, Point p2) {
        long dx = p1.x - p2.x;
        long dy = p1.y - p2.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        points = new Point[N + 1];
        parent = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            parent[i] = i;
        }

        for (int i = 1; i <= N; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            points[i] = new Point(x, y);
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            union(u, v);
        }

        List<Edge> edges = new ArrayList<>();
        for (int i = 1; i <= N; i++) {
            for (int j = i + 1; j <= N; j++) {
                double dist = distance(points[i], points[j]);
                edges.add(new Edge(i, j, dist));
            }
        }

        Collections.sort(edges);

        double answer = 0;
        for (Edge edge : edges) {
            if (union(edge.u, edge.v)) {
                answer += edge.cost;
            }
        }

        System.out.printf("%.2f\n", answer);
    }
}
